File: Day139_2025/Q139_GFG.java
import java.util.ArrayList;

class Q139_GFG {
    Node pre = null;
    Node succ = null;
    int ele1 = Integer.MIN_VALUE;
    int ele2 = Integer.MAX_VALUE;

    void findPre(Node root, int key) {
        if (root == null) return;
        if (root.data < key) {
            if (ele1 < root.data) {
                ele1 = root.data;
                pre = root;
            }
            findPre(root.right, key);
        } else {
            findPre(root.left, key);
        }
    }

    void findSucc(Node root, int key) {
        if (root == null) return;
        if (root.data > key) {
            if (ele2 > root.data) {
                ele2 = root.data;
                succ = root;
            }
            findSucc(root.left, key);
        } else {
            findSucc(root.right, key);
        }
    }

    public ArrayList<Node> findPreSuc(Node root, int key) {
        findPre(root, key);
        findSucc(root, key);
        ArrayList<Node> res = new ArrayList<>();
        res.add(pre);
        res.add(succ);
        return res;
    }
}

File: Day139_2025/Q139_GFG.py
class Solution:
    def __init__(self):
        self.pre = None
        self.succ = None
        self.ele1 = float('-inf')
        self.ele2 = float('inf')

    def findPre(self, root, key):
        if root is None:
            return
        if root.data < key:
            if self.ele1 < root.data:
                self.ele1 = root.data
                self.pre = root
            self.findPre(root.right, key)
        else:
            self.findPre(root.left, key)

    def findSucc(self, root, key):
        if root is None:
            return
        if root.data > key:
            if self.ele2 > root.data:
                self.ele2 = root.data
                self.succ = root
            self.findSucc(root.left, key)
        else:
            self.findSucc(root.right, key)

    def findPreSuc(self, root, key):
        self.findPre(root, key)
        self.findSucc(root, key)
        return [self.pre, self.succ]
